code/game/machinery/computer/security.dm

```diff
@@ -2,311 +2,192 @@
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
+/obj/machinery/computer/secure_data/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
 	if(!ui)
-		ui = new(user, src, ui_key, "secure_data.tmpl", name, 800, 800)
+		ui = new(user, src, ui_key, "SecurityRecords", name, 800, 800)
 		ui.open()
-
-/obj/machinery/computer/secure_data/ui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.default_state)
-	var/data[0]
-	data["temp"] = temp
-	data["scan"] = scan ? scan.name : null
-	data["authenticated"] = authenticated
-	data["screen"] = screen
-	if(authenticated)
-		switch(screen)
+		ui.set_autoupdate(FALSE)
+
+/obj/machinery/computer/secure_data/tgui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.default_state)
+	var/list/data = list()
+	data["currentPage"] = current_page
+	data["isPrinting"] = is_printing
+	tgui_login_data(data, user)
+	data["modal"] = tgui_modal_data()
+	data["temp"] = temp_notice
+	if(data["loginState"]["logged_in"])
+		switch(current_page)
 			if(SEC_DATA_R_LIST)
-				if(!isnull(GLOB.data_core.general))
-					for(var/datum/data/record/R in sortRecord(GLOB.data_core.general, sortBy, order))
-						var/crimstat = "null"
-						for(var/datum/data/record/E in GLOB.data_core.security)
-							if(E.fields["name"] == R.fields["name"] && E.fields["id"] == R.fields["id"])
-								crimstat = E.fields["criminal"]
-								break
-						var/background = "''"
-						switch(crimstat)
-							if(SEC_RECORD_STATUS_EXECUTE)
-								background = "'background-color:#5E0A1A'"
-							if(SEC_RECORD_STATUS_ARREST)
-								background = "'background-color:#890E26'"
-							if(SEC_RECORD_STATUS_SEARCH)
-								background = "'background-color:#999900'"
-							if(SEC_RECORD_STATUS_MONITOR)
-								background = "'background-color:#004C99'"
-							if(SEC_RECORD_STATUS_DEMOTE)
-								background = "'background-color:#C2A111'"
-							if(SEC_RECORD_STATUS_INCARCERATED)
-								background = "'background-color:#743B03'"
-							if(SEC_RECORD_STATUS_PAROLLED)
-								background = "'background-color:#743B03'"
-							if(SEC_RECORD_STATUS_RELEASED)
-								background = "'background-color:#216489'"
-							if(SEC_RECORD_STATUS_NONE)
-								background = "'background-color:#007f47'"
-							if("null")
-								crimstat = "No record."
-						data["records"] += list(list("ref" = "\ref[R]", "id" = R.fields["id"], "name" = R.fields["name"], "rank" = R.fields["rank"], "fingerprint" = R.fields["fingerprint"], "background" = background, "crimstat" = crimstat))
+				// Prepare the list of security records to associate with the general ones.
+				// This is not ideal but datacore code sucks and needs to be rewritten.
+				var/list/sec_records_assoc = list()
+				for(var/datum/data/record/S in (GLOB.data_core.security || list()))
```

Comment:
Can be made typeless.
The `|| list()` part is also not needed since `GLOB.data_core.security` should always have a value

Reply:
Has to be typeless because it's accessing `fields` inside. And OK

Reply:
By typeless I meant to not filter on the list:
```dm
for(var/thing in GLOB.data_core.security)
  var/datum/data/record/S = thing
```
If you define a type in the for loop it will filter. This list will only have records and nothing more so this is not required